### DISCUSSION TITLE
How to check columns for UNIQUE constraint

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#2437] [Q&A] [answered]
Is there a way using something like `table.columnsByName` to check if any column in the table has a UNIQUE constraint? I could use `table.uniqueKeys.isNotEmpty` but I want to also know the name of the column that has the unique constraint

**Comment by simolus3** [comment, marked as answer]
Sometimes, a unique key is not applied to a single column but rather a combination of multiple columns whose combination must be unique in the table. Those are the columns in `table.uniqueKeys` - the getter represents unique keys as a set of columns.

The information of whether a single column is unique on its own is not directly accessible from the generated models. You can get a (mediocre) approximation with something like this I guess:

```dart
bool isUnique(GeneratedColumn column) {
  final context = GenerationContext.fromDb(this);
  column.writeColumnDefinition(context);

  return context.sql.contains('UNIQUE');
}
```

But since the column constraints are only used by the migrator, the generated code is in a compact form that doesn't preserve individual columns.

**Reply by BananaMasterz** [reply]
ah I see... why do you say this is an approximation ? and a mediocre at that? Is there a chance using this code that I'll miss a unique column?

**Reply by simolus3** [reply]
I think it's unlikely that you'll miss a unique column, but if there's a column with a definition that somehow includes `UNIQUE` without it being a unique column, you'll have a false positive. That could happen with something like `text().default(Constant('UNIQUE'))()` I guess.

**Reply by BananaMasterz** [reply]
Got it. I tried the code above and it doesn't seem to work. I printed `context.sql` and it just gives the column's name.

**Reply by simolus3** [reply]
Argh, my bad! It should be `writeColumnDefinition` instead of `writeInto`.

**Reply by BananaMasterz** [reply]
ah ye, it works with `writeColumnDefinition` , ill mark this as an answer even tho it has the caveat you mentioned since in my case false positives aren't an issue. Thanks!
